cmd/vacuum/publisher.go

```diff
@@ -154,27 +209,114 @@ func getGVR(mapper meta.RESTMapper, avk *kubearchiveapi.APIVersionKind) (schema.
+		if nsRule.Name != "" {
+			// Rule has a name, it is an override
+			if idx, found := ruleNames[nsRule.Name]; found {
+				// If namespace count is lower, update the rule
+				if nsRule.Count < unifiedRules[idx].Count {
+					oldCount := unifiedRules[idx].Count
+					unifiedRules[idx].Count = nsRule.Count
+					slog.Info("Override rule count is lower, updating", "namespace", namespace, "kind", kind, "name", nsRule.Name, "newCount", nsRule.Count, "oldCount", oldCount)
+				} else {
+					slog.Error("Override rule count is not lower than cluster rule, ignoring", "namespace", namespace, "kind", kind, "name", nsRule.Name, "namespaceCount", nsRule.Count, "clusterCount", unifiedRules[idx].Count)
+				}
+			} else {
+				slog.Error("Override rule not found in cluster rules, ignoring", "namespace", namespace, "kind", kind, "name", nsRule.Name)
+			}
+		} else {
+			if celTexts[nsRule.WhenText] {
+				slog.Error("Duplicate keepLastWhen rule detected, ignoring namespace rule", "namespace", namespace, "kind", kind, "when", nsRule.WhenText)
+			} else {
+				unifiedRules = append(unifiedRules, nsRule)
+				celTexts[nsRule.WhenText] = true
+			}
+		}
 	}
 
-	if (clusterExists && kcel.ExecuteBooleanCEL(context.Background(), clusterCel.DeleteWhen, resource)) ||
-		(namespaceExists && kcel.ExecuteBooleanCEL(context.Background(), namespaceCel.DeleteWhen, resource)) {
-		return "delete-when"
+	return unifiedRules
+}
+
+func (vcep *VacuumCloudEventPublisher) createKeepers(clusterCel filters.CelExpressions, namespaceCel filters.CelExpressions, namespace string, kind string) []*Keeper {
+	unifiedRules := vcep.unifyRules(clusterCel.KeepLastWhen, namespaceCel.KeepLastWhen, namespace, kind)
+
+	keepers := make([]*Keeper, 0, len(unifiedRules))
+	for i := range unifiedRules {
+		keeper := &Keeper{
+			when:   &unifiedRules[i],
+			bucket: make([]*unstructured.Unstructured, unifiedRules[i].Count+1),
+		}
+		keepers = append(keepers, keeper)
 	}
 
-	if (clusterExists && kcel.ExecuteBooleanCEL(context.Background(), clusterCel.ArchiveWhen, resource)) ||
-		(namespaceExists && kcel.ExecuteBooleanCEL(context.Background(), namespaceCel.ArchiveWhen, resource)) {
-		return "archive-when"
+	return keepers
+}
+
+func (vcep *VacuumCloudEventPublisher) sendCloudEvent(ctx context.Context, eventType string, avk *kubearchiveapi.APIVersionKind, namespace string, name string, item *unstructured.Unstructured) {
+	sendResult := vcep.publisher.Send(ctx, eventType, item.Object)
+	var httpResult *cehttp.Result
+	statusCode := 0
+	if ce.ResultAs(sendResult, &httpResult) {
+		statusCode = httpResult.StatusCode
+	}
+	var msg string
+	if ce.IsACK(sendResult) {
+		msg = "Event sent successfully"
+	} else {
+		msg = "Event send failed"
+	}
+	slog.Info(msg, "apiversion", avk.APIVersion, "kind", avk.Kind, "namespace", namespace, "name", name, "eventType", eventType, "code", statusCode)
+}
+
+func (vcep *VacuumCloudEventPublisher) sortResources(sortField string, bucket []*unstructured.Unstructured) {
```

Comment:
Here we are sorting resources, I think we can go with this for now but, if the buckets are too large and we get OOM, we might consider an "Insert sorted" instead of appending the resource and sorting it afterwards.

A way of doing this efficiently is:

1. Pick the sorting value of the resource that is going to be inserted
2. Calculate in what index should be placed with https://pkg.go.dev/slices#BinarySearchFunc
3. Create a new slice with something like:

```go
	bucket = append(bucket, 0) // 0 is a placeholder here
	copy(bucket[idx+1:], bucket[idx:])
	bucket[idx] = newElement
```

Reply:
I'm solving this so it doesn't block the PR